ntoskrnl/ke/amd64/traphandler.c

```diff
@@ -166,25 +166,29 @@ KiSystemCallHandler(
     }
 
     /* Get descriptor table */
-    DescriptorTable = (PVOID)((ULONG_PTR)Thread->ServiceTable + Offset);
+    DescriptorTable = &((PKSERVICE_TABLE_DESCRIPTOR)Thread->ServiceTable)[TableIndex];
 
     /* Validate the system call number */
     if (ServiceNumber >= DescriptorTable->Limit)
     {
-        /* Check if this is a GUI call */
-        if (!(Offset & SERVICE_TABLE_TEST))
+        /* Check if this is a GUI call and this is not a GUI thread yet */
+        if ((TableIndex == WIN32K_SERVICE_INDEX) &&
+            (Thread->ServiceTable == KeServiceDescriptorTable))
```

Comment:
No, it's to provide the correct failure code. If you do a syscall with a win32k id that is too large, it would first attempt to convert to a gui thread, then try again, try again to convert, then fail with STATUS_ALREADY_WIN32 (in fact this is converted to STATUS_NO_MEMORY, because the former is never expected). But this call should fail with STATUS_INVALID_SYSTEM_SERVICE.